File: app/src/main/java/com/quranapp/android/adapters/ADPProphets.java
package com.quranapp.android.adapters;

import android.content.Context;
import android.content.Intent;
import android.content.res.ColorStateList;
import android.graphics.Typeface;
import android.text.SpannableString;
import android.text.TextUtils;
import android.text.style.TextAppearanceSpan;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.RecyclerView;
import static android.text.Spanned.SPAN_EXCLUSIVE_EXCLUSIVE;
import static android.view.ViewGroup.LayoutParams.WRAP_CONTENT;

import com.peacedesign.android.utils.Dimen;
import com.peacedesign.android.utils.span.LineHeightSpan2;
import com.quranapp.android.R;
import com.quranapp.android.activities.ActivityReference;
import com.quranapp.android.components.quran.QuranProphet;
import com.quranapp.android.databinding.LytQuranProphetItemBinding;
import com.quranapp.android.utils.extensions.ContextKt;
import com.quranapp.android.utils.reader.factory.ReaderFactory;

import java.text.MessageFormat;
import java.util.LinkedList;
import java.util.List;

public class ADPProphets extends RecyclerView.Adapter<ADPProphets.VHProphet> {
    private final int mItemWidth;
    private final int mTxtSize;
    private final ColorStateList mTxtColor2;
    private List<QuranProphet.Prophet> mTopics = new LinkedList<>();
    private final int mLimit;

    public ADPProphets(Context ctx, int itemWidth, int limit) {
        mItemWidth = itemWidth;
        mLimit = limit;

        mTxtSize = ContextKt.getDimenPx(ctx, R.dimen.dmnCommonSize2);
        mTxtColor2 = ContextCompat.getColorStateList(ctx, R.color.colorText2);
    }

    @Override
    public int getItemCount() {
        return mLimit > 0 ? Math.min(mTopics.size(), mLimit) : mTopics.size();
    }

    public void setProphets(List<QuranProphet.Prophet> topics) {
        mTopics = topics;
    }

    public List<QuranProphet.Prophet> getTopics() {
        return mTopics;
    }

    @NonNull
    @Override
    public VHProphet onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        LytQuranProphetItemBinding binding = LytQuranProphetItemBinding.inflate(inflater, parent, false);
        return new VHProphet(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull VHProphet holder, int position) {
        holder.bind(mTopics.get(position));
    }

    private CharSequence prepareTexts(String title, CharSequence subTitle, String inChapters) {
        SpannableString titleSS = new SpannableString(title);
        TextAppearanceSpan titleTASpan = new TextAppearanceSpan("sans-serif", Typeface.BOLD, mTxtSize, null, null);
        titleSS.setSpan(titleTASpan, 0, titleSS.length(), SPAN_EXCLUSIVE_EXCLUSIVE);

        SpannableString subTitleSS = new SpannableString(subTitle);
        TextAppearanceSpan subTitleTASpan = new TextAppearanceSpan("sans-serif", Typeface.NORMAL, mTxtSize, null, null);
        subTitleSS.setSpan(subTitleTASpan, 0, subTitleSS.length(), SPAN_EXCLUSIVE_EXCLUSIVE);
        subTitleSS.setSpan(new LineHeightSpan2(20, false, true), 0, subTitleSS.length(), SPAN_EXCLUSIVE_EXCLUSIVE);

        SpannableString chaptersSS = new SpannableString(inChapters);
        TextAppearanceSpan inChaptersTASpan = new TextAppearanceSpan("sans-serif-light", Typeface.NORMAL, mTxtSize,
            mTxtColor2,
            null);
        chaptersSS.setSpan(inChaptersTASpan, 0, chaptersSS.length(), SPAN_EXCLUSIVE_EXCLUSIVE);

        return TextUtils.concat(titleSS, "\n", subTitleSS, "\n", chaptersSS);
    }

    class VHProphet extends RecyclerView.ViewHolder {
        private final LytQuranProphetItemBinding mBinding;

        public VHProphet(@NonNull LytQuranProphetItemBinding binding) {
            super(binding.getRoot());
            mBinding = binding;

            View root = binding.getRoot();
            root.setElevation(Dimen.dp2px(root.getContext(), 4));
            root.setLayoutParams(new ViewGroup.LayoutParams(mItemWidth, WRAP_CONTENT));

            if (mItemWidth > 0) {
                root.setMinimumHeight(Dimen.dp2px(root.getContext(), 110));
                root.setBackgroundResource(R.drawable.dr_bg_chapter_card_bordered);
            } else {
                root.setBackgroundResource(R.drawable.dr_bg_chapter_card);
            }
        }

        public void bind(QuranProphet.Prophet prophet) {
            setupActions(prophet);

           /* mBinding.name.setText(MessageFormat.format("{0} ({1})", prophet.nameTrans, prophet.honorific));
            mBinding.nameEng.setText(nameEng);
            mBinding.chapters.setText(prophet.inChapters);*/

            mBinding.icon.setImageDrawable(ContextKt.drawable(mBinding.getRoot().getContext(), prophet.iconRes));

            String name = MessageFormat.format("{0} ({1})", prophet.nameTrans, prophet.honorific);
            String nameEng = "English : " + (prophet.nameEn == null ? prophet.nameTrans : prophet.nameEn);
            mBinding.text.setText(prepareTexts(name, nameEng, prophet.inChapters));
        }

        private void setupActions(QuranProphet.Prophet prophet) {
            String title = prophet.nameTrans;
            if (prophet.order == 25) {
                title += " ﷺ";
            } else {
                title += " ؑ ";
            }

            Context ctx = mBinding.getRoot().getContext();
            String desc = ctx.getString(R.string.strMsgReferenceFoundPlaces, title, prophet.verses.size());

            title = ctx.getString(R.string.strMsgReferenceInQuran, title);

            Intent intent = ReaderFactory.prepareReferenceVerseIntent(true, title, desc, new String[]{},
                prophet.chapters, prophet.verses);
            intent.setClass(ctx, ActivityReference.class);
            mBinding.getRoot().setOnClickListener(v -> ctx.startActivity(intent));
        }
    }
}

File: app/src/main/java/com/quranapp/android/adapters/ADPProphets.kt
package com.quranapp.android.adapters

import android.content.Context
import android.graphics.Typeface
import android.text.SpannableString
import android.text.Spanned
import android.text.TextUtils
import android.text.style.TextAppearanceSpan
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import com.peacedesign.android.utils.Dimen
import com.peacedesign.android.utils.span.LineHeightSpan2
import com.quranapp.android.R
import com.quranapp.android.activities.reference.ActivityReference
import com.quranapp.android.components.quran.QuranProphet
import com.quranapp.android.databinding.LytQuranProphetItemBinding
import com.quranapp.android.utils.extensions.drawable
import com.quranapp.android.utils.extensions.getDimenPx
import com.quranapp.android.utils.reader.factory.ReaderFactory.prepareReferenceVerseIntent
import com.quranapp.android.utils.sharedPrefs.SPAppConfigs
import java.text.MessageFormat
import java.util.LinkedList

class ADPProphets(ctx: Context, private val itemWidth: Int, private val limit: Int) :
    RecyclerView.Adapter<ADPProphets.VHProphet>() {
    private val txtSize = ctx.getDimenPx(R.dimen.dmnCommonSize2)
    private val txtColor2 = ContextCompat.getColorStateList(ctx, R.color.colorText2)
    var prophets: List<QuranProphet.Prophet> = LinkedList()

    override fun getItemCount(): Int {
        return if (limit > 0) prophets.size.coerceAtMost(limit) else prophets.size
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VHProphet {
        val inflater = LayoutInflater.from(parent.context)
        val binding = LytQuranProphetItemBinding.inflate(inflater, parent, false)
        return VHProphet(binding)
    }

    override fun onBindViewHolder(holder: VHProphet, position: Int) {
        holder.bind(prophets[position])
    }

    private fun prepareTexts(title: String, subTitle: CharSequence?, inChapters: String?): CharSequence {
        val titleSS = SpannableString(title).apply {
            setSpan(TextAppearanceSpan("sans-serif", Typeface.BOLD, txtSize, null, null), 0, length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
        }

        var result: CharSequence = titleSS

        if (!subTitle.isNullOrEmpty()) {
            val subTitleSS = SpannableString(subTitle).apply {
                setSpan(TextAppearanceSpan("sans-serif", Typeface.NORMAL, txtSize, null, null), 0, length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
                setSpan(LineHeightSpan2(20, false, true), 0, length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
            }
            result = TextUtils.concat(result, "\n", subTitleSS)
        }

        if (!inChapters.isNullOrEmpty()) {
            val chaptersSS = SpannableString(inChapters).apply {
                setSpan(TextAppearanceSpan("sans-serif-light", Typeface.NORMAL, txtSize, txtColor2, null), 0, length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
            }
            result = TextUtils.concat(result, "\n", chaptersSS)
        }

        return result
    }

    inner class VHProphet(private val binding: LytQuranProphetItemBinding) : RecyclerView.ViewHolder(binding.root) {
        init {
            binding.root.let {
                it.elevation = Dimen.dp2px(it.context, 4f).toFloat()
                it.layoutParams = ViewGroup.LayoutParams(itemWidth, ViewGroup.LayoutParams.WRAP_CONTENT)

                if (itemWidth > 0) {
                    it.minimumHeight = Dimen.dp2px(it.context, 110f)
                    it.setBackgroundResource(R.drawable.dr_bg_chapter_card_bordered)
                } else {
                    it.setBackgroundResource(R.drawable.dr_bg_chapter_card)
                }
            }

        }

        fun bind(prophet: QuranProphet.Prophet) {
            setupActions(prophet)

            binding.icon.setImageDrawable(binding.root.context.drawable(prophet.iconRes))

            val context = binding.root.context
            val name = MessageFormat.format("{0} ({1})", prophet.name, prophet.honorific)

            val locale = SPAppConfigs.getLocale(context)
            val isEnglish = locale == "en" || (locale == SPAppConfigs.LOCALE_DEFAULT && context.resources.configuration.locale.language == "en")

            val nameEng = if (!isEnglish || prophet.name.equals(prophet.nameEn, ignoreCase = true)) {
                null
            } else {
                "English : " + prophet.nameEn
            }

            binding.text.text = prepareTexts(name, nameEng, prophet.inChapters)
        }

        private fun setupActions(prophet: QuranProphet.Prophet) {
            val ctx = binding.root.context

            val title = ctx.getString(
                R.string.strMsgReferenceInQuran,
                MessageFormat.format("{0} ({1})", prophet.name, prophet.honorific)
            )
            val desc = ctx.getString(R.string.strMsgReferenceFoundPlaces, title, prophet.verses.size)
            val intent = prepareReferenceVerseIntent(
                true,
                title,
                desc,
                arrayOf(),
                prophet.chapters,
                prophet.verses
            )

            intent.setClass(ctx, ActivityReference::class.java)
            binding.root.setOnClickListener { ctx.startActivity(intent) }
        }
    }
}
